cdc/cdc.cc

```diff
@@ -194,4 +195,194 @@ future<> setup(service::storage_proxy& proxy, schema_ptr s) {
+        auto& t = m.token();
+        auto&& eps = _replication_strategy.get().calculate_natural_endpoints(t, _token_metadata.get());
+        if (eps.empty()) {
+            throw std::runtime_error(format("No primary replica found for key {}", m.decorated_key()));
+        }
+        auto shard_id = dht::murmur3_partitioner::shard_of(
+                t, locator::i_endpoint_snitch::get_local_snitch_ptr()->get_shard_count(eps[0]), _ignore_msb_bits);
+        mutation res(_log_schema, stream_id(*_log_schema, eps[0].addr(), shard_id));
+        set_pk_columns(m.key(), 0, res);
+        return res;
+    }
+};
+
+class streams_builder {
+    schema_ptr _schema;
+    transformer::streams_type _streams{transformer::map_cmp{}};
+    net::inet_address _node_ip = net::inet_address();
+    unsigned int _shard_id = 0;
+    api::timestamp_type _latest_row_timestamp = api::min_timestamp;
+    utils::UUID _latest_row_stream_id = utils::UUID();
+public:
+    streams_builder(schema_ptr s) : _schema(std::move(s)) {}
+
+    void accept_new_partition(const partition_key& key, uint32_t row_count) {
+        auto exploded = key.explode(*_schema);
+        _node_ip = value_cast<net::inet_address>(inet_addr_type->deserialize(exploded[0]));
+        _shard_id = static_cast<unsigned int>(value_cast<int>(int32_type->deserialize(exploded[1])));
+        _latest_row_timestamp = api::min_timestamp;
+        _latest_row_stream_id = utils::UUID();
+    }
+
+    void accept_new_partition(uint32_t row_count) {
+        assert(false);
+    }
+
+    void accept_new_row(
+            const clustering_key& key,
+            const query::result_row_view& static_row,
+            const query::result_row_view& row) {
+        auto row_iterator = row.iterator();
+        api::timestamp_type timestamp = value_cast<db_clock::time_point>(
+                timestamp_type->deserialize(key.explode(*_schema)[0])).time_since_epoch().count();
+        if (timestamp <= _latest_row_timestamp) {
+            return;
+        }
+        _latest_row_timestamp = timestamp;
+        for (auto&& cdef : _schema->regular_columns()) {
+            if (cdef.name_as_text() != "stream_id") {
+                row_iterator.skip(cdef);
+                continue;
+            }
+            auto val_opt = row_iterator.next_atomic_cell();
+            assert(val_opt);
+            val_opt->value().with_linearized([&] (bytes_view bv) {
+                _latest_row_stream_id = value_cast<utils::UUID>(uuid_type->deserialize(bv));
+            });
+        }
+    }
+
+    void accept_new_row(const query::result_row_view& static_row, const query::result_row_view& row) {
+        assert(false);
+    }
+
+    void accept_partition_end(const query::result_row_view& static_row) {
+        _streams.emplace(std::make_pair(_node_ip, _shard_id), _latest_row_stream_id);
+    }
+
+    transformer::streams_type build() {
+        return std::move(_streams);
+    }
+};
+
+static future<transformer> get_transformer(
```

Comment:
suggestion: the meat of this function is building the streams. The futurization of `transformer` is artificial (and forces it to take the schema by `schema_ptr` instead of `const schema&`, by the way). Maybe `get_streams` would be better?

Reply:
I don't understand why you say that futurization is artificial? There's a query happening inside.